### ISSUE TITLE
📎 markdown support

### ISSUE THREAD

**Issue by AugustinMauroy** [biomejs/biome#3718]
### Description

We need a specification/roadmap to enable contributors to come and contribute correctly.

* ## Parser :
For the parser I don't think we need to reinvent the wheel, we can use [`wooorm/markdown-rs`](https://github.com/wooorm/markdown-rs)
* ## Formating :
We should follow [common mark](https://commonmark.org) spec to format correctly

**Comment by ematipico** [comment]
@AugustinMauroy what's this task for? I don't really understand. Can you please put more description or context into the issue? 

Unless your really want to work on integrating markdown in Biome, we don't usually create tasks unless someone is already interested or working on it.

**Comment by AugustinMauroy** [comment]
> Unless you really want to work on integrating markdown in Biome, we don't usually create tasks unless someone is already interested or working on it.

I'd like to try, but I'm no expert on rust, so that's why I'm asking for the spec/roadmap of markdown support.

**Comment by ematipico** [comment]
We have a contribution guide that should give you the necessary information to start adding a parser inside Biome: https://github.com/biomejs/biome/blob/main/crates/biome_parser/CONTRIBUTING.md

Regarding the parsing, I believe we could use that crate, **however** we still need to map their AST to our CST because we need to own the grammar, and the internal APIs. 

We've never done this before because all our parsers are built from scratch, so it's something you should explore alone.

**Comment by suxin2017** [comment]
Recently I was digging into biome's architecture and thought I could support this

**Comment by dyc3** [comment]
I think that in order to support markdown effectively, we will need to have #3334 for formatting code blocks, but I don't think that's necessarily a hard blocker. We could start with a naive parser and formatter that doesn't format code blocks. PRs are welcome, but I expect markdown support will have the same priority as yaml support.

**Comment by AugustinMauroy** [comment]
hello ! i'm not an expert in rust etc ... But what's the next step for markdown support ?

**Comment by dyc3** [comment]
The lexer and parser is still incomplete. You could take a look in `biome_markdown_parser` and see what needs doing.

**Comment by kvz** [comment]
Would it be an option to leverage https://github.com/wooorm/markdown-rs?

**Comment by trymeouteh** [comment]
Will the Biome markdown parser also parse any HTML within a markdown file?

**Comment by dyc3** [comment]
That's the plan, but that depends on #3334

**Comment by XiNiHa** [comment]
Is MDX included in the scope? Or would that be a completely separate topic?

**Comment by ematipico** [comment]
Complete separate topic. MDX isn't a markup language, but a programming language which has its own quirks and rules that are different from markdown

**Comment by tidefield** [comment]
Hi @jfmcdowell . Given https://github.com/biomejs/biome/pull/8525 was merged (thanks!), do you have an idea what's left to claim markdown is supported?

I was thinking of implementing the parser but you beat me to it 😄  Let us know if you have further plans so that I don't step on your work.

(I'm subscribing to the issue meanwhile to be in the loop).

**Comment by jfmcdowell** [comment]
> Hi [@jfmcdowell](https://github.com/jfmcdowell) . Given [#8525](https://github.com/biomejs/biome/pull/8525) was merged (thanks!), do you have an idea what's left to claim markdown is supported?
> 
> I was thinking of implementing the parser but you beat me to it 😄 Let us know if you have further plans so that I don't step on your work.
> 
> (I'm subscribing to the issue meanwhile to be in the loop).

Hi @tidefield, for Markdown to be supported we will need a parser + formatter + linter + integrations.

#8525 started work on the parser and got to ~70% conformance to the [CommonMark spec](https://spec.commonmark.org/0.31.2/). I have local work that I am preparing for a PR that is aiming for full conformance. The remaining work is focusing on edge-case corrections and coverage.

The rest of the components need work.

The maintainers have been generous with their time and thoughtful feedback. If you’re interested in helping, I’d welcome it too.

**Comment by ematipico** [comment]
@tidefield you could lay down the foundations of the formatter e.g. `biome_markdown_formatter` with all the proper codegen/traits/etc. and the conformance test

It should be relatively trivial, but it requires some plumbing.

**Comment by tidefield** [comment]
@ematipico Yeah I can a stab at it!

**Comment by tidefield** [comment]
Some work is happening in https://github.com/biomejs/biome/pull/8962. 😉

**Comment by redtux** [comment]
> Some work is happening in [#8962](https://github.com/biomejs/biome/pull/8962). 😉

Hi @tidefield, as this was merged four days ago, can we expect markdown support to be included in the next release? Or are there other issues to be fixed before markdown can be fully supported? Thanks everybody for their help with this! 😍

**Comment by ematipico** [comment]
@redtux not really. That PR just setup the boilerplate. The parser still needs some work - but I let @jfmcdowell give the final answer as they know how much we can deem the parsing stable - and the formatter stil doesn't format anything. It just porint things verbatim 

You're more than welcome to start implementing proper formatting if you wish to get things faster to the finish line

**Comment by tidefield** [comment]
@redtux

yeah +1 to what @ematipico said. This will take some time. I also want to take the time to do this so I can understand formatters from the bottom up.

@dyc3 @ematipico I do have a question though. How do we normally go find all the specs for a new formatter? For me, I'd use AI to research https://github.com/prettier/prettier/tree/main/src/language-markdown to see what prettier supports for markdown and document a feature list. And then just tackle the list from easy to hard?

But I wonder if folks have different way to approach it.

**Comment by jfmcdowell** [comment]
> [@redtux](https://github.com/redtux) not really. That PR just setup the boilerplate. The parser still needs some work - but I let [@jfmcdowell](https://github.com/jfmcdowell) give the final answer as they know how much we can deem the parsing stable - and the formatter stil doesn't format anything. It just porint things verbatim
> 
> You're more than welcome to start implementing proper formatting if you wish to get things faster to the finish line

I feel parsing is in a good place -- with opportunities to expand to Github Flavored Markdown and as a stretch, parsing inside of code blocks. 

Here is what I think we should tackle next:

1. Formatter test harness - wire up the harness following the same patterns as the biome_css_formatter
2. Replace format_verbatim_node stubs with real formatting for block-level, code blocks, lists, and inline.
3. Workplace/CLI integration: so biome format picks them up and LSP format support
4. Markdown specific options: marker styles, bullet characters, fence style etc.
5. Lint Rules (much later)


Assuming we agree this is the work to be done I would propose:
@jfmcdowell tackle 1, 4.
@tidefield tackle 2 (block level, code blocks, lists)
@redtux if you want to help pick up 2 (inline)
TBD on 3 and all hands on deck for 5

**Comment by tidefield** [comment]
> Assuming we agree this is the work to be done I would propose:

@jfmcdowell I don't _agree_ but also don't want to disagree. For 1, what do you mean by harness exactly? For others, since we don't know what others like to work on, it's hard to be too rigid here.

Given this is volunteering-based, we don't want to be blocked or be blocker of any contributions.

**Comment by ematipico** [comment]
> For me, I'd use AI to research https://github.com/prettier/prettier/tree/main/src/language-markdown to see what prettier supports for markdown and document a feature list

Not needed, we already have infrastructure in place to copy Prettiers test and create snapshots with diffs. I believe there's a crate for that, however it requires some manual work and you need to checkout prettier repository locally and run a script. When doing so, it's advised checking out a release tag. 

The infrastructure works in way that it creates snapshots if there are diffs, and deletes them if our output is 100% identical. The infrastructure copies Prettiers tests into our repository.

**Comment by tidefield** [comment]
> When doing so, it's advised checking out a release tag

@ematipico which specific release tag should I check out? Do you mean creating a new release tag or looking at old tag as example?

**Comment by ematipico** [comment]
I mean when checking out prettier repository and copying their tests, we should doing from their latest release tag.

**Comment by tidefield** [comment]
For prettier 3.8.1, there's some bug in https://github.com/prettier/prettier/tree/3.8.1/tests/format/markdown/spec/remark-bug that breaks the script

```
Error: Cannot find module '~/prettier/tests/format/markdown/spec/remark-bug/__snapshots__/format.test.js.snap'
...
    at traverseDir (~/biome/crates/biome_formatter_test/src/prettier/prepare_tests.js:65:24)
```

This is fine that I can just skip that folder. But given https://github.com/prettier/prettier/pull/18444 and their [recent test additions](https://github.com/prettier/prettier/commits/main/tests/format/markdown) for commonmark, I wonder if we should generate the snapshots from `main` branch in this case.

It's just a coincidence that they have been working on their markdown tests recently.

**Comment by tidefield** [comment]
hmm nevermind. The script `prepare_tests.js` also relies `prettier` itselfs which is installed on my local as 3.8.1. So their recent test fixes also come with fixes inside prettier.

**Comment by jfmcdowell** [comment]
> > Assuming we agree this is the work to be done I would propose:
> 
> [@jfmcdowell](https://github.com/jfmcdowell) I don't _agree_ but also don't want to disagree. For 1, what do you mean by harness exactly? For others, since we don't know what others like to work on, it's hard to be too rigid here.
> 
> Given this is volunteering-based, we don't want to be blocked or be blocker of any contributions.

@tidefield by “test harness” I meant wiring the existing formatter test infrastructure for Markdown first, following the same pattern as CSS/JS.

Once that is in place, we can implement formatter nodes incrementally and get immediate compatibility diffs against Prettier snapshots.

**Comment by tidefield** [comment]
> I meant wiring the existing formatter test infrastructure for Markdown first

@jfmcdowell  I'm doing that in https://github.com/biomejs/biome/pull/9067

**Comment by heaths** [comment]
This might be a different issue entirely, but will we be able to associate other extensions with a markdown formatter? A static site generator uses .mdx for source material and, for the most part, they should be treated like regular markdown files.

**Comment by jfmcdowell** [comment]
> This might be a different issue entirely, but will we be able to associate other extensions with a markdown formatter? A static site generator uses .mdx for source material and, for the most part, they should be treated like regular markdown files.

IMHO full MDX support requires a dedicated MDX-aware parser (and corresponding formatter integration), not just Markdown parsing. It should be its own issue.

**Comment by ematipico** [comment]
@heaths the answer to your question was already in the thread 

https://github.com/biomejs/biome/issues/3718#issuecomment-2981091124

**Comment by heaths** [comment]
Sorry - I missed there was a collapsed section when I Cmd+F'd for "mdx". I appreciate it's not strictly markdown, but we really only want the formatting anyway. Perhaps #4754 would help, but only for formatting. I'll comment about that there.

**Comment by jfmcdowell** [comment]
@ematipico / @Netail is it worth opening an umbrella issue for tracking progress and discussion on markdown linting?

**Comment by Netail** [comment]
Mhm I don't think we really did for others (only tickets for a single rule, but that was more on request basis)

For a new language support ticket (like this one) we do usually have "adding at least 1 linting rule" in the checklist before the ticket can be marked as done

**Comment by dyc3** [comment]
Sure, a new umbrella issue for lint rules would make sense. It would give us a space to determine what is/isn't in scope.

**Comment by igor-koop** [comment]
Hey, maybe one could also consider [`bruits/satteri`](https://github.com/bruits/satteri) for the markdown processor? It supports MDAST and HAST plugins and could very organically replace and expand upon the respective Remark and Rehype plugins of the unified markdown processor for formatting and linting, for instance, by substituting [`remarkjs/remark-lint`](https://github.com/remarkjs/remark-lint). The added benefit might be that Satteri could be used as a standard processor, for example, within [Astro](https://astro.build/blog/astro-640/#new-markdownprocessor-api), and formatting and linting would be performed with the same framework used to compile the webpages, which is useful for extensions that add new syntax, such as MDX or GFM.

**Comment by guillp** [comment]
For Markdown linting, you also might want to consider [rumdl](https://github.com/rvben/rumdl) which is a superset of `markdownlint`, implemented in Rust, and ultra-fast.

**Comment by tats-u** [comment]
> we can use [wooorm/markdown-rs](https://github.com/wooorm/markdown-rs)

> Hey, maybe one could also consider [bruits/satteri](https://github.com/bruits/satteri) for the markdown processor?

They don't support the CST output, which Biome requires.

----

Biome currently doesn't support any GFM syntax including tables.

However, using Prettier's unconditional column alignment across all lines creates several issues:

- Including extremely long strings in some cells creates excessive whitespace in other rows.
- CJK characters' width isn't always exactly twice that of Latin characters (when a CJK font and a Latin monospace font are combined), resulting in alignment issues in CJK languages. This problem may extend to other Asian scripts as well.
- Modifying content in one row often propagates changes to other rows, increasing VCS changes and making conflicts more likely.
